### ISSUE TITLE
bloc: Support 2D Boolean selection and assignment in Frame

### ISSUE THREAD

**Issue by flexatone** [InvestmentSystems/static-frame#85]
This might be done by extending `loc` functionality, or by creating a new type of selector interface.

```python
In [70]: f                                                                                                                                                                  
Out[70]: 
<Frame>
<Index> a       b       <<U1>
<Index>
0       1       3
1       2       4
<int64> <int64> <int64>

In [71]: f == 2                                                                                                                                                             
Out[71]: 
<Frame>
<Index> a      b      <<U1>
<Index>
0       False  False
1       True   False
<int64> <bool> <bool>

In [72]: f.loc[f == 2]                                                                                                                                                      
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
```

**Comment by flexatone** [comment]
Possible interfaces might look like this:

`Frame.bloc(np.ndarray) -> Series`: similar to `loc` or `__getitem__` interfaces, but takes a Boolean 2D array (of matching shape, or reindexable to matching shape if a Frame) and returns a `Series` indexed by tuple pairs from the index/column axis. Returning a `Frame` will not always be possible, as a Boolean 2D array might specify values in non-contiguous regions. It does not appear that returning a sparse `Frame`, of the outermost bounding box possible  to contain the selected values, would be useful. If function application is desired on the selected values, `iter_element.apply` could be called on the returned `Series`.

`Frame.assign.bloc(np.ndarray)(value) -> Frame`: when used for assignment, a `Frame` of the same shape (and index/columns) is returned. Assignment to the Boolean values specified by the input array can be provided by either (a) a `Frame` that has values at corresponding indices (or NaN if missing, simulating reindexing), (b) a same shaped unindexed array, from which values are drawn at corresponding iloc positions, or (c) a `Series` indexed by tuple coordinates that match the values targeted from the Boolean array.

**Comment by ForeverWintr** [comment]
I came here looking for a similar feature. Here's my use case for your consideration.

I want to reformat a frame for display by replacing some values with strings. For example:

```python
f = sf.Frame([[True, False], [False, True]], index=('a', 'b'), columns=['d', 'c'])
f = f.assign[f]('T')
f = f.assign[~f]('')

f.to_tsv('/tmp/tmp.tsv)
```

This is the desired behavior in pandas:

```python
df = f.to_pandas()
df[~df] = ''
df[df==True] = 'T'
df
   d  c
a  T   
b     T
```

**Comment by flexatone** [comment]
This has been implemented as the `Frame.assign.bloc` interface, and will be released in 0.5. There is also a `Frame.bloc` selector, returning a `Series` from `Frame`.

```python
>>> import static_frame as sf
>>> f = sf.Frame([[True, False], [False, True]], index=('a', 'b'), columns=['d', 'c'])
>>> f
<Frame>
<Index> d      c      <<U1>
<Index>
a       True   False
b       False  True
<<U1>   <bool> <bool>

>>> f.assign.bloc(f)('T')
<Frame>
<Index> d        c        <<U1>
<Index>
a       T        False
b       False    T
<<U1>   <object> <object>
 
>>> f.assign.bloc(f)('T').assign.bloc(~f)('')
<Frame>
<Index> d        c        <<U1>
<Index>
a       T
b                T
<<U1>   <object> <object>
```
